modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/QueryCancel.java

```diff
@@ -32,21 +36,66 @@ public class QueryCancel {
 
     private boolean canceled;
 
+    private boolean timeout;
+
     /**
-     * Adds a cancel action.
+     * Adds a cancel action. If operation has already been canceled, throws a {@link QueryCancelledException}.
+     *
+     * <p>NOTE: If the operation is cancelled, this method will immediately invoke the given action
+     * and then throw a {@link QueryCancelledException}.
      *
-     * @param clo Add cancel action.
+     * @param clo Cancel action.
      */
     public synchronized void add(Cancellable clo) throws QueryCancelledException {
         assert clo != null;
 
         if (canceled) {
-            throw new QueryCancelledException();
+            // Immediately invoke a cancel action, if already cancelled. 
+            // Otherwise the caller is required to catch QueryCancelledException and call an action manually.
+            try {
+                clo.cancel(timeout);
+            } catch (Exception ignore) {
+                // Do nothing
+            }
+
+            String message = timeout ? QueryCancelledException.TIMEOUT_MSG : QueryCancelledException.CANCEL_MSG;
+            throw new QueryCancelledException(message);
         }
 
         cancelActions.add(clo);
     }
 
+    /**
+     * Removes the given callback.
+     *
+     * @param clo Callback.
+     */
+    public synchronized void remove(Cancellable clo) {
+        assert clo != null;
+
+        cancelActions.remove(clo);
+    }
+
+    /**
+     * Schedules a timeout action (a call to {@link #timeout()}) after {@code timeoutMillis} milliseconds.
+     *
+     * @param scheduler Scheduler to trigger an action.
+     * @param timeoutMillis Timeout in milliseconds.
+     * @return Future that will be completed when the timeout is reached.
+     */
+    public CompletableFuture<Void> addTimeout(ScheduledExecutorService scheduler, long timeoutMillis) {
+        CompletableFuture<Void> fut = new CompletableFuture<>();
+        fut.thenAccept((r) -> timeout());
+
+        ScheduledFuture<?> f = scheduler.schedule(() -> {
+            fut.complete(null);
+        }, timeoutMillis, MILLISECONDS);
+
+        add((timeout) -> f.cancel(false));
```

Comment:
Do I understand correctly that this code is needed to cancel a future when the request has completed without a timeout?

Then it might be worth adding 
```
if (!timeout) f.cancel(false)
```
because the logic for canceling a future may not be so trivial :thinking:

Reply:
> if (!timeout) f.cancel(false)
This is not correct. When a statement is cancelled its timeout should be cancelled as well, otherwise a ScheduledFuture is still alive but a query does not exist anymore.

Reply:
> When a statement is cancelled its timeout should be cancelled as well,

Yes, but when timeout occurred then this future `f` must be already completed and we don't need to cancel it again?
```
add((timeout) -> {

assert timeout == false || f.isDone(); // <-- is this correct?

f.cancel(false);

}
```

Reply:
> assert timeout == false || f.isDone(); // <-- is this correct?

These assertions are not necessary, since Future::cancel is idempotent.

`QueryCancel` expects its callbacks to be idempotent as well. It is a in multithreaded environment after all, if multiple actors want to terminate a query, only one of them can succeed (other actors should just accept the fact that someone else has already done the job they tried to do).

Reply:
Just to make this clear. For every statement we create a QueryCancel thing that stores actions to be called when a statement is cancelled.  

A statement can is cancelled when on initiator in case of an error. DistributedQueryManage calls query cancel : cancel method. QueryCancel cancel calls all cancellation callbacks.

This commit introduced a timeout method on QueryCancel to distinguish between a cancellation due to some exception and a time out.

- QueryProcessor creates an instance of QueryTimeout q.
- it schedules a timeout action for a query that calls q::timeout (to indicate that timeout occurred and a statement should be cancelled)
- when a statement is cancelled due to some exception by its initiator a scheduled future should be cancelled as well.